Source hubspot: add incremental streams

artem1205 opened the pull request:
## What
Resolving https://github.com/airbytehq/airbyte/issues/19265

## How
add cursors with filtering by date

## Recommended reading order
1. `x.java`
2. `y.python`

## 🚨 User Impact 🚨

no breaking changes

## Pre-merge Checklist


<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Connector version has been incremented
    - [ ] Version has been bumped according to our [Semantic Versioning for Connectors](https://docs.airbyte.com/contributing-to-airbyte/semantic-versioning-for-connectors) guidelines
    - [ ] `Dockerfile` has updated version
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` with an entry for the new version. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)

- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub and connector version bumped by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</details>

artem1205 commented:
/test connector=connectors/source-hubspot

lazebnyi commented on airbyte-integrations/connectors/source-hubspot/source_hubspot/streams.py:
```diff
@@ -664,6 +666,45 @@ def _flat_associations(self, records: Iterable[MutableMapping]) -> Iterable[Muta
+        record_value = (
+            pendulum.parse(record.get(self.cursor_field)).int_timestamp
+            if isinstance(record.get(self.cursor_field), str)
+            else record.get(self.cursor_field) // 1000
```
why 1000?

artem1205 replied:
refactored. date-fields  is stored in milliseconds from now.

artem1205 commented:
/test connector=connectors/source-hubspot

artem1205 commented:
/test connector=connectors/source-hubspot

artem1205 commented:
/test connector=connectors/source-hubspot

artem1205 commented:
/test connector=connectors/source-hubspot

sherifnada commented on airbyte-integrations/connectors/source-hubspot/source_hubspot/streams.py:
```diff
@@ -664,6 +666,46 @@ def _flat_associations(self, records: Iterable[MutableMapping]) -> Iterable[Muta
+    _cursor_value = 1
+
+    @property
+    def cursor_field(self) -> Union[str, List[str]]:
```
should this be also an `@abstractmethod` to force all subclasses to define this field?

artem1205 replied:
`@abstractmethod` added

sherifnada commented on airbyte-integrations/connectors/source-hubspot/source_hubspot/streams.py:
```diff
@@ -664,6 +666,46 @@ def _flat_associations(self, records: Iterable[MutableMapping]) -> Iterable[Muta
+        record_value, datetime_format = (
+            (pendulum.parse(record.get(self.cursor_field)), "YYYY-MM-DDTHH:mm:ss.SSSSSSZ")
+            if isinstance(record.get(self.cursor_field), str)
+            else (pendulum.from_format(str(record.get(self.cursor_field)), "x"), "x")
```
what is `x`?  also can you add a code comment about why we have this `if` statement?

artem1205 replied:
docsting added

sherifnada replied:
this is a smaller comment but fwiw this should be done using inversion of control i.e: the super class should not know what the subclasses do. Instead the subclasses should have a way to provide the datetime object given their knowledge of their own date format.

sherifnada reviewed (changes requested):
As this is a GA connector, it's super important to overcommunicate the customer impact in both breaking changes and docs. Some specific points:

1. Why do SATs have a breaking change when the PR description says there's no breaking change? Please leave that as a source code comment
2. The external facing docs should be updated to describe this change
3. there should be unit tests for this new semi incremental stream
4. there are some code comments to address as well

artem1205 commented:
/test connector=connectors/source-hubspot

artem1205 commented on airbyte-integrations/connectors/source-hubspot/source_hubspot/streams.py:
```diff
@@ -664,6 +666,46 @@ def _flat_associations(self, records: Iterable[MutableMapping]) -> Iterable[Muta
 
 
+class SemiIncrementalStream(Stream, IncrementalMixin):
+    _cursor_value = 1
```
changed to empty string to avoid confusion, no magic  :smile:

artem1205 commented:
/test connector=connectors/source-hubspot

sherifnada commented on airbyte-integrations/connectors/source-hubspot/unit_tests/test_streams.py:
```diff
@@ -75,33 +75,33 @@ def test_updated_at_field_non_exist_handler(requests_mock, common_params, fake_p
     ],
 )
-def test_streams_read(stream, endpoint, requests_mock, common_params, fake_properties_list):
+def test_streams_read(stream, endpoint, cursor_value, requests_mock, common_params, fake_properties_list):
```
does this test that the filtering works properly? at the `SemiIncrementalStream` level? I think no, right?

sherifnada commented on airbyte-integrations/connectors/source-hubspot/source_hubspot/streams.py:
```diff
@@ -664,6 +666,57 @@ def _flat_associations(self, records: Iterable[MutableMapping]) -> Iterable[Muta
             yield record
 
 
+class SemiIncrementalStream(Stream, IncrementalMixin):
```
nit: a better name for this pattern is `ClientSideIncrementalStream`

artem1205 replied:
renamed

sherifnada commented on docs/integrations/sources/hubspot.md:
please add a note to this doc that there are two kinds of incremental sync: server-side incremental sync (i.e: where the hubspot API returns only the data updated or generated since the last sync) and client-side incremental sync (where the Hubspot API returns all the data / doesn't allow filtering and the connector filters only the records which were updated) and update the `Supported Streams` list to incidcate which sync types they support

artem1205 replied:
docs updated

artem1205 commented:
@sherifnada , agreed with inversion of control principle. I refactored and added `@abstractmethod` for `cursor_field_datetime_format`.
Unittest for ClientSideIncrementalStream also added

artem1205 commented:
/test connector=connectors/source-hubspot

erohmensing commented on airbyte-integrations/connectors/source-hubspot/sample_files/incremental_catalog.json:
```diff
@@ -237,23 +240,17 @@
       "destination_sync_mode": "append"
     },
-    {
-      "stream": {
-        "name": "ticket_pipelines",
-        "json_schema": {},
-        "supported_sync_modes": ["full_refresh"]
-      },
-      "sync_mode": "full_refresh",
-      "destination_sync_mode": "overwrite"
-    },
```
Why not include this stream in the incremental catalog example?

artem1205 replied:
this stream is empty, while for incremental tests, all streams should produce at least 1 record

erohmensing commented on airbyte-integrations/connectors/source-hubspot/sample_files/incremental_catalog.json:
```diff
@@ -154,19 +160,13 @@
-      "destination_sync_mode": "overwrite"
-    },
-    {
-      "stream": {
-        "name": "form_submissions",
-        "json_schema": {},
-        "supported_sync_modes": ["full_refresh"]
```
Why not include this stream in the incremental catalog example/acceptance test?

artem1205 replied:
this stream is empty, while for incremental tests, all streams should produce at least 1 record

erohmensing replied:
And we don't have `empty_streams` for incremental tests, got it. Thanks for the explanation!

erohmensing commented on airbyte-integrations/connectors/source-hubspot/Dockerfile:
```diff
@@ -34,5 +34,5 @@ COPY source_hubspot ./source_hubspot
 ENTRYPOINT ["python", "/airbyte/integration_code/main.py"]
 
-LABEL io.airbyte.version=0.4.0
+LABEL io.airbyte.version=0.5.0
```
Backwards compatible new features: minor release 👌🏻

erohmensing reviewed (approved):
Reviewing from the connector ops side: 
* Changes to acceptance-test-config.yml and sample files used in it
* Version bumping, semver, changelog

artem1205 commented:
/publish connector=connectors/source-hubspot
